### PR TITLE
Feature/nex 52/resource tree e2e tests

### PR DESCRIPTION
**Pull request by oatymart**
https://oat-sa.atlassian.net/browse/NEX-52

Ready for review. Please checkout the branch and also install https://www.npmjs.com/package/@oat-sa/tao-e2e-runner into your `tao/views/build` folder.
The tests should appear when you open the Cypress app.

There is one test (creating a subclass) which is currently skipped because the behaviour of the BRS is unexpected (it closes the tree after renaming a class). I could force this test to pass with a few extra commands, but it seems better to fix the bug first.

TODO
- ~~make tests pass~~
- ~~test against empty and full Resource Tree~~
- ~~improve commands with parent-child style~~
- ~~version bump~~

### CODE REVIEW

File: views/js/e2e/resourceSelector/item/item.spec.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        it('can create and rename a new item', function() {
+            cy.addItem(selectors.itemsRootClass);
+
+            cy.renameSelectedItem(newItemName);
+
+            cy.get(selectors.resourceTree)
+                .contains(newItemName)
+                .should('exist')
+                // .and('be.visible');
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/item/item.spec.js]: ?

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+Cypress.Commands.add('loadItemsPage', () => {
+    cy.fixture('urls')
+    .as('urls')
+    .then(urls => {
+        // Provide the full URL parameters including 'uri'
+        // to guarantee a predictable tree with the 'Item' root class selected
+        cy.visit(`${urls.index}?${urls.taoItemsRoot}&${urls.nosplashParam}`);
+        // Important to register this first response, or it will mess up future "wait"s:
+        cy.wait('@editClass');
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: In my docker stack the load takes a bit more time sometimes then 5 secs, we should consider to add a bit of timeout

File: views/js/e2e/resourceSelector/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# About these tests
+
+The tests in this folder are for the **resourceSelector** component in TAO CE.
+
+Please note that the resourceSelector looks and behaves differently from the **jsTree** component which is loaded by default in (e.g.) TAO 3.3.0-sprint101.
+
+To install the resourceSelector, run the following command from your instance root:
+
+```
+php index.php "\\oat\\tao\\scripts\\install\\SetResourceSelector"
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/README.md]: Command not working in docker environment, we have to find an alternative way to set this.

File: views/js/e2e/resourceSelector/class/class.spec.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                    .should('exist');
+            });
+        });
+
+        it('has correct action buttons when class is selected', function() {
+            cy.selectTreeNode(selectors.itemsRootClass);
+
+            // check the visible action buttons
+            cy.get(selectors.actionsContainer).within(() => {
+                Cypress._.forEach([
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/class/class.spec.js]: Proper DOM selectors pls, id (why do we have ids on elements???) or title data or class which for sure does not have a translation.

File: views/js/e2e/resourceSelector/item/item.spec.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+            cy.renameSelectedItem(newItemName);
+
+            cy.get(`[title="${newItemName}"]`)
+                .closest(selectors.treeNode)
+                .should('not.have.class', 'closed');
+
+            cy.get(selectors.actionsContainer).within(() => {
+                Cypress._.forEach([
+                    'New class',
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/item/item.spec.js]: Proper DOM selectors pls, id or title data or class which for sure does not have a translation

File: views/js/e2e/resourceSelector/item/item.spec.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+        it('has correct action buttons when nothing is selected', function() {
+            // deselect selected list item
+            cy.get(selectors.resourceTree)
+                .find('.selected').first()
+                .click({ force: true });
+
+            cy.get(selectors.actionsContainer).within(() => {
+                Cypress._.forEach([
+                    'New class',
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/item/item.spec.js]: Proper DOM selectors pls, id or title  or class which for sure does not have a translation

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    // this event needs to fire twice before proceeding
+    cy.wait('@editClass').wait('@editClass').wait(300);
+});
+
+Cypress.Commands.add('renameSelectedItem', (newName) => {
+    cy.log('COMMAND: renameSelectedItem', newName);
+
+    // assumes that editing form has already been rendered
+    cy.get(selectors.contentContainer).within(() => {
+        cy.contains('label', 'Label')
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: dom selector pls for this element

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            .siblings('input')
+            .should('be.visible')
+            .clear()
+            .type(newName);
+
+        cy.contains('Save')
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: dom selector pls for this element

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    // this event needs to fire twice before proceeding
+    cy.wait(['@editItem', '@editItem']).wait(300);
+});
+
+Cypress.Commands.add('addClass', (cssSelector) => {
+    cy.log('COMMAND: addClass', cssSelector);
+
+    cy.selectTreeNode(cssSelector);
+
+    cy.contains('New class').click();
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: dom selector pls for this element

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    // this event needs to fire twice before proceeding
+    cy.wait('@editClass').wait('@editClass');
+});
+
+Cypress.Commands.add('addItem', (cssSelector) => {
+    cy.log('COMMAND: addItem', cssSelector);
+
+    cy.selectTreeNode(cssSelector);
+
+    cy.contains('New item').click();
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: dom selector pls for this element

File: views/js/e2e/resourceSelector/resourceTree.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+            .siblings('input')
+            .should('be.visible')
+            .clear()
+            .type(newName);
+
+        cy.contains('Save')
+            .click();
+    });
+    // this event needs to fire twice before proceeding
+    cy.wait('@editClass').wait('@editClass').wait(300);
```
Comment (ferenckiss89) [inline views/js/e2e/resourceSelector/resourceTree.js]: why is this 300 ms needed?

File: views/js/e2e/resourceSelector/item/item.spec.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            cy.addItem(selectors.itemsRootClass);
+
+            cy.renameSelectedItem(newItemName);
+
+            cy.get(`[title="${newItemName}"]`)
+                .closest(selectors.treeNode)
+                .should('not.have.class', 'closed');
+
+            cy.get(selectors.actionsContainer).within(() => {
+                Cypress._.forEach([
```
Comment (btamas) [inline views/js/e2e/resourceSelector/item/item.spec.js]: I'm not a fan of lodash especially if the functionality totally supported natively.

```
['newClass', ...].forEach(action => {...})
```

**Pull request merged**